File: regions/system/library/go/building-blocks/statestore.go
package buildingblocks

import "context"

// ETag represents an optimistic concurrency token.
type ETag struct {
	Value string `json:"value"`
}

// StateEntry represents a stored state value.
type StateEntry struct {
	Key   string `json:"key"`
	Value []byte `json:"value"`
	ETag  *ETag  `json:"etag,omitempty"`
}

// SetRequest represents a request to set state.
type SetRequest struct {
	Key   string `json:"key"`
	Value []byte `json:"value"`
	ETag  *ETag  `json:"etag,omitempty"`
}

// StateStore provides state management with optimistic concurrency via ETags.
type StateStore interface {
	Component
	Get(ctx context.Context, key string) (*StateEntry, error)
	Set(ctx context.Context, req *SetRequest) (*ETag, error)
	Delete(ctx context.Context, key string, etag *ETag) error
	BulkGet(ctx context.Context, keys []string) ([]*StateEntry, error)
	BulkSet(ctx context.Context, requests []*SetRequest) ([]*ETag, error)
}

File: regions/system/library/typescript/building-blocks/src/statestore.ts
import type { Component } from './component.js';

export interface StateEntry {
  key: string;
  value: Uint8Array;
  etag: string;
}

export interface StateStore extends Component {
  get(key: string): Promise<StateEntry | null>;
  set(key: string, value: Uint8Array, etag?: string): Promise<string>;
  delete(key: string, etag?: string): Promise<void>;
  bulkGet(keys: string[]): Promise<StateEntry[]>;
  bulkSet(entries: Array<{ key: string; value: Uint8Array }>): Promise<string[]>;
}
